cpp/src/rolling/detail/optimized_unbounded_window.cpp

```diff
@@ -0,0 +1,155 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/groupby/sort_helper.hpp>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/groupby.hpp>
+#include <cudf/reduction/detail/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/unary.hpp>
+#include <cudf/utilities/default_stream.hpp>
+
+namespace cudf::detail {
+
+bool can_optimize_unbounded_window(bool unbounded_preceding,
+                                   bool unbounded_following,
+                                   size_type min_periods,
+                                   rolling_aggregation const& agg)
+{
+  auto is_supported = [](auto const& agg) {
+    switch (agg.kind) {
+      case cudf::aggregation::Kind::COUNT_ALL:
+      case cudf::aggregation::Kind::COUNT_VALID:
+      case cudf::aggregation::Kind::SUM:
+      case cudf::aggregation::Kind::MIN:
+      case cudf::aggregation::Kind::MAX: return true;
+      default:
+        // COLLECT_LIST and COLLECT_SET can be added at a later date.
+        // Other aggregations do not fit into the [UNBOUNDED, UNBOUNDED]
+        // category. For instance:
+        // 1. Ranking functions (ROW_NUMBER, RANK, DENSE_RANK, PERCENT_RANK)
+        //    use [UNBOUNDED PRECEDING, CURRENT ROW].
+        // 2. LEAD/LAG are defined on finite row boundaries.
+        return false;
+    }
+  };
+
+  return unbounded_preceding && unbounded_following && (min_periods == 1) && is_supported(agg);
+}
+
+/// Converts rolling_aggregation to corresponding reduce/groupby_aggregation.
+template <typename Base>
+struct aggregation_converter {
+  template <aggregation::Kind k>
+  std::unique_ptr<Base> operator()() const
+  {
+    if constexpr (std::is_same_v<Base, cudf::groupby_aggregation> and
+                  k == aggregation::Kind::COUNT_ALL) {
+      // Note: cudf::reduce() does not support COUNT_ALL.
+      return cudf::make_count_aggregation<Base>(null_policy::INCLUDE);
+    } else if constexpr (std::is_same_v<Base, cudf::groupby_aggregation> and
+                         k == aggregation::Kind::COUNT_VALID) {
+      // Note: cudf::reduce() does not support COUNT_VALID.
+      return cudf::make_count_aggregation<Base>(null_policy::EXCLUDE);
+    } else if constexpr (k == aggregation::Kind::SUM) {
+      return cudf::make_sum_aggregation<Base>();
+    } else if constexpr (k == aggregation::Kind::MIN) {
+      return cudf::make_min_aggregation<Base>();
+    } else if constexpr (k == aggregation::Kind::MAX) {
+      return cudf::make_max_aggregation<Base>();
+    } else {
+      CUDF_FAIL("Unsupported aggregation kind: " + std::to_string(static_cast<int>(k)));
```

Comment:
The other errors I can see of this type do not report the aggregation. I don't think we get much from having an integer here, anyway, and I don't know if there's an easy string conversion.
```suggestion
      CUDF_FAIL("Unsupported aggregation kind.");
```